### PR TITLE
Set the user agent on all http calls and update for SmartSync to support MergeModeOptions on SyncUp.

### PR DESCRIPTION
**Pull request by smcnulty-sfdc**

### CODE REVIEW

**Comment by khawkins** [comment]
:+1:

File: SalesforceSDK/Salesforce.SDK.Core/Source/Net/HttpCall.cs [hunk 1]
```diff
@@ -89,15 +91,20 @@ public static string MimeType(this ContentTypeValues contentType)
...
-        private readonly string Url;
-        private readonly HttpClient WebClient;
-        private Exception HttpCallErrorException;
-        private string ResponseBodyText;
-        private HttpStatusCode StatusCodeValue;
+        /// <summary>
+        /// Use this property to retrieve the user agent.
+        /// </summary>
+        public static string UserAgentHeader { private set; get; }
+        private const string UserAgentHeaderFormat = "SalesforceMobileSDK/3.1 {0}";
```
Comment (wmathurin) [inline SalesforceSDK/Salesforce.SDK.Core/Source/Net/HttpCall.cs]: Where is the rest of it (see https://github.com/forcedotcom/SalesforceMobileSDK-Android/blob/master/libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.java#L911)

Comment (smcnulty-sfdc) [inline SalesforceSDK/Salesforce.SDK.Core/Source/Net/HttpCall.cs]: Added application name and application version to the UserAgent.

Example UserAgent:

SalesforceMobileSDK/3.1 (Salesforce.SDK.Store.Test/1.1.0 native) 5.0 (Windows NT 6.3; WOW64; Trident/7.0; Touch; .NET4.0E; .NET4.0C; Tablet PC 2.0; .NET CLR 3.5.30729; .NET CLR 2.0.50727; .NET CLR 3.0.30729; WebView/2.0; rv:11.0) like Gecko

File: SalesforceSDK/Salesforce.SDK.Core/Source/Net/HttpCall.cs [hunk 2]
```diff
@@ -273,43 +285,54 @@ public static HttpCall CreatePost(string url, string requestBody)
...
-            if (!String.IsNullOrWhiteSpace(RequestBody))
+            // if the user agent has not yet been set, set it; we want to make sure this only really happens once since it requires an action that goes to the core thread.
+            if (String.IsNullOrWhiteSpace(UserAgentHeader))
+            {
+                CoreDispatcher core = CoreApplication.MainView.CoreWindow.Dispatcher;
+                await core.RunAsync(CoreDispatcherPriority.Normal, async () => await GenerateOsString());
+                var packageVersion = Package.Current.Id.Version;
+                var packageVersionString = packageVersion.Major + "." + packageVersion.Minor + "." +
+                                           packageVersion.Build;
+                UserAgentHeader = String.Format(UserAgentHeaderFormat, await GetApplicationDisplayName(), packageVersionString, "native", UserAgentHeader);
```
Comment (bhariharan) [inline SalesforceSDK/Salesforce.SDK.Core/Source/Net/HttpCall.cs]: Format used in Android:

```String.format("SalesforceMobileSDK/%s android mobile/%s (%s) %s/%s %s", SDK_VERSION, Build.VERSION.RELEASE, Build.MODEL, appName, appVersion, nativeOrHybrid)```

Comment (bhariharan) [inline SalesforceSDK/Salesforce.SDK.Core/Source/Net/HttpCall.cs]: Based on offline discussion with @smcnulty-sfdc, most of this stuff is not exposed in ```Windows```.

**Comment by bhariharan** [comment]
Looks good to me!

File: SalesforceSDK/Salesforce.SDK.Core/Source/Net/HttpCall.cs (hunk 2 above)
Comment (wmathurin) [inline SalesforceSDK/Salesforce.SDK.Core/Source/Net/HttpCall.cs]: You might want to send an example of a salesforce mobile sdk windows user agent to the mobile services team to see if their user agent parsing code will correctly handle it.

**Comment by wmathurin** [comment]
Looks fine.

**Pull request merged**